File: backend/app/schemas/market.py
"""Market data, news, and technicals schemas."""

from datetime import datetime

from pydantic import BaseModel


# ---- Bars ----

class BarOut(BaseModel):
    timestamp: datetime
    open: float
    high: float
    low: float
    close: float
    volume: int
    vwap: float | None = None

    class Config:
        from_attributes = True


class PriceHistoryOut(BaseModel):
    ticker: str
    bars: list[BarOut]
    bar_count: int
    period_start: str | None = None
    period_end: str | None = None
    period_change_pct: float | None = None
    latest_close: float | None = None


# ---- News ----

class ArticleOut(BaseModel):
    id: str
    ticker: str
    published_at: datetime
    source: str | None = None
    headline: str
    summary: str | None = None
    url: str | None = None
    relevance_score: int = 0

    class Config:
        from_attributes = True


class NewsListOut(BaseModel):
    ticker: str
    article_count: int
    articles: list[ArticleOut]


# ---- Technicals ----

class TechnicalsOut(BaseModel):
    ticker: str
    latest_close: float | None = None
    sma_5: float | None = None
    sma_10: float | None = None
    sma_20: float | None = None
    rsi_14: float | None = None
    latest_volume: int | None = None
    avg_volume_20: int | None = None
    volume_vs_avg_pct: float | None = None
    period_high: float | None = None
    period_low: float | None = None
    bars_analyzed: int | None = None
    error: str | None = None

File: frontend/src/lib/types/market.ts
// Types that mirror backend Pydantic schemas.
// Keep these in sync with backend/app/schemas/*.

export type WatchlistItem = {
  ticker: string;
  position: number;
  added_at: string;
};

export type Bar = {
  timestamp: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
  vwap?: number | null;
};

export type PriceHistory = {
  ticker: string;
  bars: Bar[];
  bar_count: number;
  period_start?: string | null;
  period_end?: string | null;
  period_change_pct?: number | null;
  latest_close?: number | null;
};

export type Article = {
  id: string;
  ticker: string;
  published_at: string;
  source?: string | null;
  headline: string;
  summary?: string | null;
  url?: string | null;
  relevance_score: number;
};

export type NewsList = {
  ticker: string;
  article_count: number;
  articles: Article[];
};

export type Technicals = {
  ticker: string;
  latest_close?: number | null;
  sma_5?: number | null;
  sma_10?: number | null;
  sma_20?: number | null;
  rsi_14?: number | null;
  latest_volume?: number | null;
  avg_volume_20?: number | null;
  volume_vs_avg_pct?: number | null;
  period_high?: number | null;
  period_low?: number | null;
  bars_analyzed?: number | null;
  error?: string | null;
};
